crates/ruma-identity-service-api/src/invitation/store_invitation/v2.rs

```diff
@@ -0,0 +1,112 @@
+//! [POST /_matrix/identity/v2/store-invite](https://matrix.org/docs/spec/identity_service/r0.3.0#post-matrix-identity-v2-store-invite)
+
+use ruma_api::ruma_api;
+use ruma_common::thirdparty::Medium;
+use ruma_identifiers::{MxcUri, RoomAliasId, RoomId, UserId};
+
+ruma_api! {
+    metadata: {
+        description: "Store pending invitations to a user's 3PID.",
+        method: POST,
+        name: "store_invitation",
+        path: "/_matrix/identity/v2/store-invite",
+        authentication: AccessToken,
+        rate_limited: false,
+    }
+
+    request: {
+        /// The type of the third party identifier for the invited user.
+        ///
+        /// Currently, only `Medium::Email` is supported.
+        pub medium: &'a Medium,
+
+        /// The email address of the invited user.
+        pub address: &'a str,
+
+        /// The Matrix room ID to which the user is invited.
+        pub room_id: &'a RoomId,
+
+        /// The Matrix user ID of the inviting user.
+        pub sender: &'a UserId,
+
+        /// The Matrix room alias for the room to which the user is invited.
+        ///
+        /// This should be retrieved from the `m.room.canonical` state event.
+        #[serde(skip_serializing_if = "Option::is_none")]
+        pub room_alias: Option<&'a RoomAliasId>,
+
+        /// The Content URI for the room to which the user is invited.
+        ///
+        /// This should be retrieved from the `m.room.avatar` state event.
+        #[serde(skip_serializing_if = "Option::is_none")]
+        pub room_avatar_url: Option<&'a MxcUri>,
+
+        /// The `join_rule` for the room to which the user is invited.
+        ///
+        /// This should be retrieved from the `m.room.join_rules` state event.
+        #[serde(skip_serializing_if = "Option::is_none")]
+        pub room_join_rules: Option<&'a str>,
+
+        /// The name of the room to which the user is invited.
+        ///
+        /// This should be retrieved from the `m.room.name` state event.
+        #[serde(skip_serializing_if = "Option::is_none")]
+        pub room_name: Option<&'a str>,
+
+        /// The display name of the user ID initiating the invite.
+        #[serde(skip_serializing_if = "Option::is_none")]
+        pub sender_display_name: Option<&'a str>,
+
+        /// The Content URI for the avater of the user ID initiating the invite.
+        #[serde(skip_serializing_if = "Option::is_none")]
+        pub sender_avatar_url: Option<&'a MxcUri>,
+    }
+
+    response: {
+        /// The generated token. Must be a string consisting of the characters `[0-9a-zA-Z.=_-]`.
+        ///
+        /// Its length must not exceed 255 characters and it must not be empty.
+        pub token: String,
+
+        /// A list of [server's long-term public key, generated ephemeral public key].
+        pub public_keys: Vec<String>,
```

Comment:
It looks like this is basically a tuple: can you investigate whether we can deserialize it as one?

I think it might be nice to be able to do:

```rust
let server_key = response.public_keys.server_key;
let ephemeral_key = response.public_key;

// do some stuff...
```
```rust
let server_key = response.public_keys.get(0).ok_or_else( /* ... */ );
let ephemeral_key = response.public_keys.get(1).ok_or_else( /* ... */ );

// do some stuff...
```

Then again, this is pretty concise, so maybe we don't need it?
```rust
if let [server_key, ephemeral_key] = &response.public_keys[0..2] {
    // do some stuff
}
```

Reply:
I think your first code snippet is missing `s.ephemeral_key` before the second semicolon. Otherwise I like it. There are (De)Serialize impls for fixed-size arrays from 0 to 32 so those can be used to implement `Deserialize`:

```rust
let array = [String; 2]::deserialize(deserializer)?;
// turn array into struct PublicKeys
```

while `Serialize` can be implemented using `serialize_seq` (https://docs.rs/serde/1.0.126/serde/ser/trait.SerializeSeq.html#example-use).

Reply:
> I think your first code snippet is missing s.ephemeral_key before the second semicolon.

Fixed, thanks.